### DISCUSSION TITLE
Intrinsic parameter for o3r camera

### DISCUSSION THREAD

**Discussion by Astronomia-maker** [ifm/ifm3d#392] [Q&A]
Hi,

How can i get the intrinsic parameter of camera head (O3R 225) for python environment?

**Comment by dekhanra** [comment]
Hi @Astronomia-maker,

The intrinsic parameters  of the camera head are the part of TOF INFO chunk which can be extracted by de-serializing the data returned from buffer_id TOF_INFO.

Install the latest unstable ifm3dpy test version from testPyPI:
```shell
$ pip install -i https://test.pypi.org/simple/ ifm3dpy
```
In the new ifm3dpy version the deserialization is implemented as method. The following code snippet may help you.

```python
from ifm3dpy.device import O3R
from ifm3dpy.deserialize import TOFInfoV4 as deserialize_tof
from ifm3dpy.framegrabber import FrameGrabber, buffer_id
import time

o3r = O3R()
fg = FrameGrabber(o3r, pcic_port=50012) # Expected 3D head connected to PORT2

#set schema and start Grabber
fg.start([buffer_id.TOF_INFO])
[ok, frame] = fg.wait_for_frame().wait_for(1500) # wait with 1500ms timeout

if ok:
    tof_info_buffer = frame.get_buffer(buffer_id.TOF_INFO)
    tof_info_buffer = deserialize_tof.deserialize(tof_info_buffer)

    intrinsic_calib = tof_info_buffer.intrinsic_calibration
    print(intrinsic_calib.parameters)


fg.stop()
```

**Reply by dekhanra** [reply]
You can browse the documentation from [https://api.ifm3d.com/latest/](https://api.ifm3d.com/latest/)

**Reply by dekhanra** [reply]
For more information regarding the calibrations please look at https://github.com/ifm/o3r-utilities/tree/main/intro_to_calibrations

**Reply by Astronomia-maker** [reply]
Hi @dekhanra, Thank you for code snippet. After running your code i am getting 
`[128.57142639160156,128.57142639160156, 109.99063873291016, 85.79098510742188, 0.0, 0.502928970.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]        9722442627, -0.40336400270462036, 0.0, 0.0, 0.6367549896240234, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,.0 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]`  this output. So is there any description available for better understading?

**Reply by dekhanra** [reply]
Hi @Astronomia-maker,

I guess the length of output array should be 32 may be a Copy-Paste mistake. However, the modelID of the camera head will be 2 in my guess(As you have mentioned O3R225) but can you recheck with (`intrinsic_calib.model_id`) in print statement in code snippet.

Therefore the first 10 entries of the modelParameters array refers to 

**f<sub>x</sub>, f<sub>y</sub>, m<sub>x</sub>, m<sub>y</sub>, $\alpha$, k<sub>1</sub>, k<sub>2</sub>, k<sub>3</sub>, k<sub>4</sub>, $\theta$<sub>max</sub>**

**Reply by Astronomia-maker** [reply]
Hi @dekhanra, 
I have added another print statement for `(intrinsic_calib.model_id)` and have this output `[128.57142639160156, 128.57142639160156, 109.9906387329111, 85.63045501708984, 0.0, 0.50292016, 85.79098510742188, 0.0, 0.5029289722442627, -0.40330, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.6400270462036, 0.0, 0.0, 0.6367549896240234, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
0`

**Reply by dekhanra** [reply]
Hi @Astronomia-maker ,

This script holds all the information regarding the Intrinsic Calibration parameters:  https://github.com/ifm/o3r-utilities/blob/main/registration2d3d/transforms.py
